staging/src/k8s.io/apimachinery/pkg/runtime/converter.go

```diff
@@ -317,80 +415,141 @@ func sliceFromUnstructured(sv, dv reflect.Value) error {
-			return err
+		origLen := ctx.pushIndex(i)
+		se, err := fromUnstructured(sv.Index(i), dv.Index(i), ctx)
+		if err != nil {
+			return nil, err
 		}
+		if se != nil {
+			strictErrs = ctx.appendStrictErrors(strictErrs, se...)
+		}
+		ctx.popPath(origLen)
 	}
-	return nil
+	return strictErrs, nil
+}
+
+func getSeparator(path []string) string {
+	if len(path) > 0 {
+		return "."
+	}
+	return ""
 }
 
-func pointerFromUnstructured(sv, dv reflect.Value) error {
+func pointerFromUnstructured(sv, dv reflect.Value, ctx *fromUnstructuredContext) (strictErrs []error, err error) {
 	st, dt := sv.Type(), dv.Type()
 
 	if st.Kind() == reflect.Ptr && sv.IsNil() {
 		dv.Set(reflect.Zero(dt))
-		return nil
+		return nil, nil
 	}
 	dv.Set(reflect.New(dt.Elem()))
 	switch st.Kind() {
 	case reflect.Ptr, reflect.Interface:
-		return fromUnstructured(sv.Elem(), dv.Elem())
+		return fromUnstructured(sv.Elem(), dv.Elem(), ctx)
 	default:
-		return fromUnstructured(sv, dv.Elem())
+		return fromUnstructured(sv, dv.Elem(), ctx)
 	}
 }
 
-func structFromUnstructured(sv, dv reflect.Value) error {
+func structFromUnstructured(sv, dv reflect.Value, ctx *fromUnstructuredContext) (strictErrs []error, err error) {
 	st, dt := sv.Type(), dv.Type()
 	if st.Kind() != reflect.Map {
-		return fmt.Errorf("cannot restore struct from: %v", st.Kind())
+		return nil, fmt.Errorf("cannot restore struct from: %v", st.Kind())
 	}
 
+	defer func() {
+		ctx.parentPath = ctx.parentPath[:len(ctx.parentPath)]
+	}()
 	for i := 0; i < dt.NumField(); i++ {
 		fieldInfo := fieldInfoFromField(dt, i)
 		fv := dv.Field(i)
 
 		if len(fieldInfo.name) == 0 {
-			// This field is inlined.
-			if err := fromUnstructured(sv, fv); err != nil {
-				return err
+			// This field is inlined, recurse into fromUnstructured again
+			// with the same set of matched keys.
+			origInlined := ctx.isInlined
+			ctx.isInlined = true
+			se, err := fromUnstructured(sv, fv, ctx)
+			if err != nil {
+				return nil, err
+			}
+			if se != nil {
+				strictErrs = append(strictErrs, se...)
 			}
+			ctx.isInlined = origInlined
 		} else {
+			// This field is not inlined so we recurse into
+			// child field of sv corresponding to field i of
+			// dv, with a new set of matchedKeys and updating
+			// the parentPath to indicate that we are one level
+			// deeper.
+			ctx.addMatchedKey(fieldInfo.name)
 			value := unwrapInterface(sv.MapIndex(fieldInfo.nameValue))
 			if value.IsValid() {
-				if err := fromUnstructured(value, fv); err != nil {
-					return err
+				origLen := ctx.pushKey(fieldInfo.name)
+				origInlined := ctx.isInlined
+				origMatchedKeys := ctx.matchedKeys
+				ctx.isInlined = false
+				ctx.matchedKeys = nil
+				se, err := fromUnstructured(value, fv, ctx)
+				if err != nil {
+					return nil, err
 				}
+				if se != nil {
+					strictErrs = append(strictErrs, se...)
+				}
+				ctx.popPath(origLen)
+				ctx.isInlined = origInlined
+				ctx.matchedKeys = origMatchedKeys
```

Comment:
It feels a lot like your'e actually calling `fromUnstructured` with a new context, except for the path. Can you have `pushKey` return a new context? You'd need to figure out a way to re-use the same allocated path when you pop (possibly using a pointer to the path).

Then you could possibly go as far as doing:
```suggestion
				se, err := fromUnstructured(value, fv, ctx.WithPath(fieldInfo.name))
				if err != nil {
					return nil, err
				}
				if se != nil {
					strictErrs = append(strictErrs, se...)
				}
```

Reply:
If I'm remembering correctly, I think Jordan explicitly did not want us to have the parent path as a pointer (but I can't find the original convo about this). Doing so would force us to do a bunch of ugly dereferencing every where we use the path and it seems a bit more brittle because we would have to add in a bunch of nil checks in places where ctx.parenthPath is being used to avoid NPEs.

Is your concern here just that the stashing of original fields is ugly or do you feel like it effects correctness/performance in some way? If it's the latter, can I suggest instead just using helpers to stash and restore the context like I've added here? I think any gains we get from being able to neatly create a new context as you suggest will be negated by the ugliness of trying to re-use the same allocated path and stashing/restore the non-path fields of the context I would argue feels sligthly simpler overall.

I agree about putting the length in the context and have done so now.